go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -698,29 +699,44 @@ func (npw *nodePortWatcher) DeleteEndpoints(ep *kapi.Endpoints) {
 	}
 }
 
-func (npw *nodePortWatcher) UpdateEndpoints(old *kapi.Endpoints, new *kapi.Endpoints) {
-	name := ktypes.NamespacedName{Namespace: old.Namespace, Name: old.Name}
+func getEndpointAddresses(endpointSlice *discovery.EndpointSlice) []string {
+	endpointsAddress := make([]string, 0)
+	for _, endpoint := range endpointSlice.Endpoints {
+		endpointsAddress = append(endpointsAddress, endpoint.Addresses...)
+	}
+	return endpointsAddress
+}
 
-	if reflect.DeepEqual(new.Subsets, old.Subsets) {
+func (npw *nodePortWatcher) UpdateEndpointSlice(oldEpSlice, newEpSlice *discovery.EndpointSlice) {
+	oldEpAddr := getEndpointAddresses(oldEpSlice)
+	newEpAddr := getEndpointAddresses(newEpSlice)
+	if reflect.DeepEqual(oldEpAddr, newEpAddr) {
 		return
 	}
 
-	klog.V(5).Infof("Updating endpoints %s in namespace %s", old.Name, old.Namespace)
-
+	namespacedName := namespacedNameFromEPSlice(oldEpSlice)
+	// This case is to handle when we update from externalName service
+	// to nodeport service, update endpointslice handler is called and
+	// to have the nodeport IPtable rules installed.
+	if _, exists := npw.getServiceInfo(namespacedName); !exists {
```

Comment:
@ricky-rav this check condition was the extra thing that was added form the previous MR and few extra mock test code.
In case of endpoints when we update service from ExternalName to nodeport and create a corresponding endpoint for nodeport, Addendpoints handler is called and nodeport iptables are populated as expected where as with endpointslice when we update from ExternalName service to nodeport service and when corresponding endpoint is added **UpdateEndpoinstlice** handler function is called out and UpdateEndpointslice just check whether there is any diff between old and endpoint slice which isn't and the nodeport Ip table rules are being missed out as a result.

The extra check that I included is to check if the service details are already present in service cache (which get updated normally when a cluster IP service or nodeport service gets added) so in this case of extername service service details get cached and so I check if that is scenario, I just call the Addendpointslice ahndler for this update case (externalName service to nodeport service update) and so nodeport Iptabe rule gets added. for other updates the cache will alwasy be present and AddEndpointslice handler wont be called thrugh updateEndpiointslice.